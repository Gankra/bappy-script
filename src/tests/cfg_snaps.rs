fn dump_cfg(input: &str) -> String {
    let mut program = crate::Program::typed(input);
    program.parse().unwrap();
    program.check();

    let ctx = program.ctx.unwrap();
    let cfg = program.cfg.unwrap();

    cfg.format(&ctx).unwrap()
}

#[test]
fn capture_basics() {
    let program = r#"
    struct Point {
        x: Int,
        y: Int,
    }
    
    let pt = Point { x: 2, y: 7 }
    fn captures(arg: Int) -> Int {
        ret add(arg, pt.y)
    }

    fn no_captures(arg: Int) -> Int {
        ret add(arg, arg)
    }

    fn calls_funcs() -> Int {
        let a = captures(23)
        let b = no_captures(29)
        ret add(a, b)
    }

    let my_fn = captures
    let my_fn2 = no_captures

    print captures(2)
    print my_fn(3)

    print no_captures(6)
    print my_fn2(7)

    print calls_funcs()

    let x = 1
    let y = true
    let z = (x, y)
    if z.1 {
        print add(1, add(x, 2))
    }
    ret 0
"#;

    let cfg = dump_cfg(program);
    insta::assert_snapshot!("basics", cfg);
}

#[test]
fn loop_basics() {
    let program = r#"
    let cond = true
    let mut counter = 3
    loop {
        if eq(counter, 0) {
            break
        }
        set counter = sub(counter, 1)
        print counter

        loop {
            print "hello"
            if cond {
                break
            } else {
                continue
            }
            print "unreachable"
        }
        print "looped!"
    }
    ret 0
"#;

    let cfg = dump_cfg(program);
    insta::assert_snapshot!("loop_basics", cfg);
}

#[test]
fn func_names_are_paths() {
    let program = r#"
    fn do_a_compy(x: Int) -> Int {
        ret mul(x, x)
    }
    let z = 5
    fn captured_do_a_compy(x: Int) -> Int {
        ret do_a_compy(add(x, z))
    }
    struct MyClosure {
        func: fn(Int) -> Int
        capture: Int
    }
    fn call_closure(close: MyClosure) -> Int {
        ret close.func(close.capture)
    }
    
    let close1 = MyClosure { func: do_a_compy, capture: 7 }
    let close2 = MyClosure { func: captured_do_a_compy, capture: 9 }
    
    print call_closure(close1)
    print call_closure(close2)
    
    ret close2.capture
    "#;

    let cfg = dump_cfg(program);
    insta::assert_snapshot!("func_names_are_paths", cfg);
}
