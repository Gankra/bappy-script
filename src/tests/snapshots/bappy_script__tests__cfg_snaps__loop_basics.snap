---
source: src/tests/cfg_snaps.rs
assertion_line: 86
expression: cfg

---
[intrinsic] #fn0_add(lhs, rhs)
[intrinsic] #fn1_sub(lhs, rhs)
[intrinsic] #fn2_mul(lhs, rhs)
[intrinsic] #fn3_eq(lhs, rhs)
[intrinsic] #fn4_not(rhs)

#fn5_main():
  bb0():
    %0 = Bool(true)
    %1 = Int(3)
    %2 = stack_alloc() (slot 0)
    %2.* = %1
    jmp bb1()

  bb1():
    %3 = %2.*
    %4 = Int(0)
    %5 = #fn3_eq(%3, %4)
    cond %5: bb3(), bb4()

  bb2():
    %13 = Int(0)
    ret %13

  bb3():
    jmp bb2()
    jmp bb5()

  bb4():
    jmp bb5()

  bb5():
    %6 = %2.*
    %7 = Int(1)
    %8 = #fn1_sub(%6, %7)
    %2.* = %8
    %9 = %2.*
    print %9
    jmp bb6()

  bb6():
    %10 = Str("hello")
    print %10
    cond %0: bb8(), bb9()

  bb7():
    %12 = Str("looped!")
    print %12
    jmp bb1()

  bb8():
    jmp bb7()
    jmp bb10()

  bb9():
    jmp bb6()
    jmp bb10()

  bb10():
    %11 = Str("unreachable")
    print %11
    jmp bb6()


